### PR TITLE
Refactor storage of poll data

### PR DESCRIPTION
**Pull request by apetrus20**

### CODE REVIEW

File: react/features/polls/components/AbstractPollAnswer.tsx [hunk 1]
```diff
@@ -50,16 +49,16 @@ const AbstractPollAnswer = (Component: ComponentType<AbstractProps>) => (props:
...
         if (poll.lastVote !== null) {
             return [ ...poll.lastVote ];
         }
 
         return new Array(poll.answers.length).fill(false);
     });
-    const participant = useBoundSelector(getParticipantById, poll.senderId);
+
+    const reduxState = useSelector(state => state);
+    const participantName = getParticipantDisplayName(reduxState, poll.senderId);
```
Comment (saghul) [inline react/features/polls/components/AbstractPollAnswer.tsx:61]: Pl write this as a selector, not using that auxiliary reduxState

File: react/features/polls/middleware.ts [hunk 2]
```diff
@@ -81,9 +80,15 @@ MiddlewareRegistry.register(({ dispatch, getState }) => next => action => {
         const { conference } = action;
 
         conference.on(JitsiConferenceEvents.ENDPOINT_MESSAGE_RECEIVED,
-            (_: any, data: any) => _handleReceivePollsMessage(data, dispatch));
+            (_: any, data: any) => {
```
Comment (saghul) [inline react/features/polls/middleware.ts:83]: PL rename the parameter instead of using _

File: react/features/polls/middleware.ts [hunk 3]
```diff
@@ -81,9 +80,15 @@ MiddlewareRegistry.register(({ dispatch, getState }) => next => action => {
...
+                data.type === COMMAND_NEW_POLL ? data.senderId = _._id : data.voterId = _._id;
+                _handleReceivePollsMessage(data, dispatch);
+            });
         conference.on(JitsiConferenceEvents.NON_PARTICIPANT_MESSAGE_RECEIVED,
-            (_: any, data: any) => _handleReceivePollsMessage(data, dispatch));
+            (_: any, data: any) => {
```
Comment (saghul) [inline react/features/polls/middleware.ts:88]: ditto

**Review by saghul — CHANGES_REQUESTED** [review]
Please take a look at the linter errors.

File: react/features/polls/components/AbstractPollAnswer.tsx [hunk 4]
```diff
@@ -5,7 +5,9 @@ import { useDispatch, useSelector } from 'react-redux';
 import { createPollEvent } from '../../analytics/AnalyticsEvents';
 import { sendAnalytics } from '../../analytics/functions';
 import { IReduxState } from '../../app/types';
-import { getLocalParticipant, getParticipantById } from '../../base/participants/functions';
+// eslint-disable-next-line lines-around-comment
+// @ts-ignore
+import { getParticipantDisplayName } from '../../base/participants';
```
Comment (saghul) [inline react/features/polls/components/AbstractPollAnswer.tsx:10]: import this from /functions and you can remove the ts-ignore.

**Review by saghul — COMMENTED** [review]
LGTM, though I haven't tested it yet. Does the mobile part need an update?

**Comment by saghul** [comment]
Update: tested and working nicely, good job! Please check mobile needs no changes and address the nit so ts-ignore is not needed and we're good to go!

**Comment by saghul** [comment]
What the status of this?

**Comment by apetrus20** [comment]
> What the status of this?

I made all the changes and I tested it on mobile as well, it's working nicely on mobile as well.

File: react/features/polls/reducer.ts [hunk 5]
```diff
@@ -83,21 +83,37 @@ ReducerRegistry.register<IPollsState>('features/polls', (state = INITIAL_STATE,
         // if the poll exists, we update it with the incoming answer
         const newAnswers = state.polls[pollId].answers
             .map(_answer => {
+                const voters = _answer.voters as string[];
+
                 return {
                     name: _answer.name,
-                    voters: new Map(_answer.voters)
+                    voters: typeof voters?.length === 'number' ? [ ...voters ] : { ...voters }
```
Comment (saghul) [inline react/features/polls/reducer.ts:90]: Can't we normalize this at this level? That is, if we get an object / map transfor it into an array?

Also, I feel like this needs a comment to make sure what's happening is properly understood.

File: react/features/polls/reducer.ts [hunk 6]
```diff
@@ -83,21 +83,37 @@ ReducerRegistry.register<IPollsState>('features/polls', (state = INITIAL_STATE,
...
-            const voters = newAnswers[i].voters;
-
-            if (answer.answers[i]) {
-                voters.set(answer.voterId, answer.voterName);
 
+        for (let i = 0; i < newAnswers.length; i++) {
+            // if the answer was chosen, we add the senderId to the array of voters of this answer
+            const voters = newAnswers[i].voters as any;
+
+            if (typeof voters?.length === 'number') {
```
Comment (saghul) [inline react/features/polls/reducer.ts:99]: Yeah all this looks odd. Let's convert the old format to the new one and then only process that please.

**Review by saghul — CHANGES_REQUESTED** [review]

File: react/features/polls/reducer.ts [hunk 7]
```diff
@@ -83,21 +83,29 @@ ReducerRegistry.register<IPollsState>('features/polls', (state = INITIAL_STATE,
         // if the poll exists, we update it with the incoming answer
         const newAnswers = state.polls[pollId].answers
             .map(_answer => {
+                const answerVoters = _answer.voters
```
Comment (saghul) [inline react/features/polls/reducer.ts:86]: Can you pl add a comment here that explains you are supporting the old data model?

**Review by saghul — APPROVED** [review]

**Pull request merged**